src/ds/ptrwrap.h

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+#include <atomic>
+
+namespace snmalloc
+{
+  template<typename T>
+  using Pointer = T*;
+
+  template<typename T>
+  using AtomicPointer = std::atomic<T*>;
+
+  /**
+   * Summaries of StrictProvenance metadata.  We abstract away the particular
+   * size and any offset into the bounds.
+   *
+   * Yes, I wish the start-of-comment characters were aligned below as well.
+   * I blame clang format.
+   */
+  enum ptrauth_bounds
+  {
+    /*                Spatial  Notes                               */
+    PAArena, /*        Arena                                       */
+    PASlab, /*          Slab                                       */
+    PASlabE, /*         Slab   (+ platform constraints)            */
+    PAAlloc, /*        Alloc                                       */
+    PAAllocE /*        Alloc   (+ platform constraints)            */
+  };
+
+  /**
+   * Compute the "exported" variant of a ptrauth_bounds annotation.  This is
+   * used by the PAL's ptrauth_export function to compute its return value's
+   * annotation.
+   */
+  template<ptrauth_bounds B>
+  constexpr ptrauth_bounds ptrauth_export_type()
+  {
+    static_assert(
+      (B == PASlab) || (B == PAAlloc), "ptrauth_export_type of bad type");
+    if constexpr (B == PASlab)
+    {
+      return PASlabE;
+    }
+    else if constexpr (B == PAAlloc)
+    {
+      return PAAllocE;
+    }
+  }
+
+  /**
+   * A pointer annotated with a "phantom type parameter" carrying a static
+   * summary of its StrictProvenance metadata.
+   */
+  template<typename T, ptrauth_bounds bounds>
+  struct AuthPtr
+  {
+    T* unsafe_auth_ptr;
+
+    /**
+     * nullptr is constructable at any bounds type
+     */
+    AuthPtr(const std::nullptr_t n) : unsafe_auth_ptr(n) {}
+
+    AuthPtr() : unsafe_auth_ptr(nullptr) {}
+
+    explicit AuthPtr(T* p) : unsafe_auth_ptr(p) {}
```

Comment:
Why does MSVC think my explicit constructor isn't reachable? :(

Reply:
Aaaargh.  So, something about the `sandbox`'s use of an allocator that never needs initialization results in MSVC getting unhappy.  In particular, it seems to correctly realize that the `large_alloc()` `NeedsInitialization` code path cannot happen, but something about that path's need to unwrap and rewrap the `AuthPtr<>` (as/from a `void*` to pass through the `InitThreadAllocator` callback) results in it generating this warning that the `AuthPtr` ctor itself is unreachable. :(

I've added the `#pragma warning(push)`-and-friends dance and documentation around the constructor itself (sigh) as part of 53c30a6342.